src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/tools/cortisol/IngestTask.java

```diff
@@ -0,0 +1,172 @@
+                // 0 indicates an interrupted exception, so break out and exit.
+                if (tokenBucket.take() != 0) {
+                    List<String> docs = new ArrayList<>();
+                    for (int i = 0; i < params.getDocsPerRequest(); ++i) {
+                        String doc = br.readLine();
+                        if (doc == null) {
+                            // restart the indexing from the top.
+                            br = new BufferedReader(new FileReader(params.getTrackLocation()));
+                            doc = br.readLine();
+                        }
+
+                        docs.add(doc);
+                    }
+                    makeBulkRequest(docs);
+                } else {
+                    break;
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private RestHighLevelClient buildEsClient() {
+        final RestClientBuilder builder = RestClient.builder(new HttpHost(endpoint, port, "http"));
+        return new RestHighLevelClient(builder);
+    }
+
+    private void createIndex(final String mappingTemplateJson, int nPrimaryShards, int nReplicaShards) throws IOException {
+        final CreateIndexRequest cir = new CreateIndexRequest(indexName);
+        cir.mapping(mappingTemplateJson, XContentType.JSON);
+        cir.settings(CortisolHelper.buildIndexSettingsJson(nPrimaryShards, nReplicaShards), XContentType.JSON);
+        final CreateIndexResponse response = client.indices().create(cir, RequestOptions.DEFAULT);
+        assert response.isAcknowledged();
+    }
+
+    private void makeBulkRequest(final List<String> bulkDocs) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (String bulkDoc : bulkDocs) {
+            final IndexRequest indexRequest = Requests.indexRequest(indexName);
+            indexRequest.source(bulkDoc, XContentType.JSON);
+            bulkRequest.add(indexRequest);
+        }
+        client.bulkAsync(bulkRequest, RequestOptions.DEFAULT, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                if (!bulkItemResponses.hasFailures()) {
+                    successCount.addAndGet(bulkDocs.size());
+                } else {
+                    for (BulkItemResponse response : bulkItemResponses) {
+                        if (response.isFailed()) {
+                            failureCount.incrementAndGet();
+                        } else {
+                            successCount.incrementAndGet();
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                failureCount.addAndGet(bulkDocs.size());
+            }
+        });
+        requestCount.addAndGet(bulkDocs.size());
+    }
+
+    private class TokenBucket {
```

Comment:
You should add a comment to this class since it functions slightly different than a standard token bucket. I believe this lets you make at most nTokens concurrent take() calls every second.

Reply:
Done.